source/code/plugins/agent_maintenance_script.rb

```diff
@@ -190,13 +196,49 @@ def apply_dsc_endpoint(server_resp)
       # When apply_dsc_endpoint is called from onboarding, dsc_endpoint will be returned in file
       update_config("DSC_ENDPOINT", dsc_endpoint)
 
+      generate_dsc_metaconfig
+
       return dsc_endpoint
     end
 
+    # Generate the DSC MetaConfig file. If the dsc endpoint is updated, DSC will
+    # not directly pick the endpoint from omsadmin.conf (instead, it is picked
+    # from MetaConfig.mof), so we must manually trigger metaconfig (re)generation.
+    def generate_dsc_metaconfig
+      # OMS_MetaConfigHelper.py will log to omsconfig.log by default,
+      command = "/opt/microsoft/omsconfig/Scripts/OMS_MetaConfigHelper.py"
+      timeout = 10
+      begin
+        Open3.popen2(command, :pgroup=>true) do |_, _, wait_thr|
+          pid = wait_thr.pid
+          pgid = Process.getpgid(pid)
+          deadline = Time.now + timeout
+          sleep 0.5 until Time.now > deadline or !wait_thr.status
+          if wait_thr.status # still hasn't completed
+            begin
+              `pkill -TERM -g #{pgid}`
+              sleep 1
+              if wait_thr.status
+                `pkill -KILL -g #{pgid}`
+              end
+            rescue => e
+              log_error("Failed to kill process(es) for command '#{command}'. #{e}")
+            end
+            OMS::Log.warn_once("DSC MetaConfig generation failed due to subprocess timeout")
```

Comment:
I have tested this and it works. I already detailed how I tested this in the PR description.

If you check the OMS_MetaConfigHelper.py it doesn't need any parameters if it is being run to generate metaconfig. The parameters are only necessary if you are trying to disable or enable DSC config management.

Reply:
OMS_MetaConfigHelper.py persists the argument/flag that it was executed with previously by creating file (/etc/opt/omi/conf/omsconfig/omshelper_disable).

Assume a user runs OMS_MetaConfigHelper.py with --disable flag. This would change the Metaconfig from PULL mode to PUSH mode. If the user runs the script second time with out any flag, then the script will still use previous argument and configures Metaconfig in PUSH mode where DSC endpoint is not configured.

We need to test the changes considering scenario where the DSC endpoint change is happened when the omsagent is in disables state and is enabled later by the user.

Reply:
I've added the check to see if DSC is disabled. If so, I won't call `OMS_MetaConfigHelper.py` but will persist a flag so once DSC is re-enabled we will call.